central/graphql/resolvers/node_vulnerabilities.go

```diff
@@ -80,8 +82,366 @@ func (resolver *Resolver) NodeVulnCounter(ctx context.Context, args RawQuery) (*
+	vulnLoader, err := loaders.GetNodeCVELoader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	query = tryUnsuppressedQuery(query)
+	fixableVulnsQuery := search.ConjunctionQuery(query, search.NewQueryBuilder().AddBools(search.Fixable, true).ProtoQuery())
+	fixableVulns, err := vulnLoader.FromQuery(ctx, fixableVulnsQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	unFixableVulnsQuery := search.ConjunctionQuery(query, search.NewQueryBuilder().AddBools(search.Fixable, false).ProtoQuery())
+	unFixableCVEs, err := vulnLoader.FromQuery(ctx, unFixableVulnsQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	return mapNodeCVEsToVulnerabilityCounter(fixableVulns, unFixableCVEs), nil
+}
+
+// CreatedAt is the time a node CVE first seen in the system
+func (resolver *nodeCVEResolver) CreatedAt(ctx context.Context) (*graphql.Time, error) {
+	value := resolver.data.GetCveBaseInfo().GetCreatedAt()
+	return timestamp(value)
+}
+
+// CVE name of the node CVE
+func (resolver *nodeCVEResolver) CVE(ctx context.Context) string {
+	return resolver.data.GetCveBaseInfo().GetCve()
+}
+
+// EnvImpact is the fraction of nodes that contain the nodeCVE
+func (resolver *nodeCVEResolver) EnvImpact(ctx context.Context) (float64, error) {
+	n, d, err := resolver.getEnvImpactComponentsForNodeCVE(ctx)
+	if err != nil {
+		return 0, err
+	}
+	if d == 0 {
+		return 0, nil
+	}
+	return float64(n) / float64(d), nil
+}
+
+func (resolver *nodeCVEResolver) getEnvImpactComponentsForNodeCVE(ctx context.Context) (numerator, denominator int, err error) {
+	allNodesCount, err := resolver.root.NodeGlobalDataStore.CountAllNodes(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+	if allNodesCount == 0 {
+		return 0, 0, nil
+	}
+	nodeLoader, err := loaders.GetNodeLoader(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+	withThisCVECount, err := nodeLoader.CountFromQuery(resolver.withNodeVulnerabilityScope(ctx), search.EmptyQuery())
+	if err != nil {
+		return 0, 0, err
+	}
+	return int(withThisCVECount), allNodesCount, nil
+}
+
+func (resolver *nodeCVEResolver) withNodeVulnerabilityScope(ctx context.Context) context.Context {
+	return scoped.Context(ctx, scoped.Scope{
+		ID:    resolver.data.GetId(),
+		Level: v1.SearchCategory_NODE_VULNERABILITIES,
+	})
+}
+
+// FixedByVersion returns the version of the parent component that removes this CVE
+func (resolver *nodeCVEResolver) FixedByVersion(ctx context.Context) (string, error) {
+	return resolver.getNodeComponentFixedByVersion(ctx)
+}
+
+func (resolver *nodeCVEResolver) getNodeComponentFixedByVersion(_ context.Context) (string, error) {
+	scope, hasScope := scoped.GetScope(resolver.ctx)
+	if !hasScope {
+		return "", nil
+	}
+	if scope.Level != v1.SearchCategory_NODE_COMPONENTS {
+		return "", nil
+	}
+	edgeID := edges.EdgeID{ParentID: scope.ID, ChildID: resolver.data.GetId()}.ToString()
+	edge, found, err := resolver.root.NodeComponentCVEEdgeDataStore.Get(resolver.ctx, edgeID)
+	if err != nil || !found {
+		return "", err
+	}
```

Comment:
`FixedByVersion` is a sub-resolver of `nodeCVEResolver`. That is, we only expect calls like,
```
nodeVulnerabilities {
  fixedByVersion
}
```
On `ln251` we ensure that ctx should contain component scope. Therefore, when node vulnerability already pulled out of db, it already has the fixedby field value correctly scoped, and hence it can it returned here.

Reply:
So we do not need to check for `found` ?

Reply:
If `scope.Level == v1.SearchCategory_NODE_COMPONENTS` we know that component scope was incorporated when node vulnerability was fetched from the store (and the scope is an inner join— `select * from node_cves inner join componentcveedge on node_cves.id == componentcveedge.nodecveid ....`. The node vuln, with the component scope, is fetched only because the edge exists. Hence, the `found` is known to be true when you are here.

if `scope.Level != v1.SearchCategory_NODE_COMPONENTS`, the code returns empty string.